In [2]:
#Import Packages: Pandas for Data Wrangling
import pandas as pd

We have two raw data files that we need to wrangle by cleaning, aggregating, and combining in order to analyze effectively. The two files include a CSV file containing Recycling Diversion & Capture rates for NYC broken down by district along with a CSV file containing district level demographic, economic, and community data. 

In order to work with the CSV files containing the data, we will begin by loading in the file through the Pandas library. Below, we will read in and preview the NYC recycling rates file.

In [3]:
#Read in and preview Recycling Diversion & Capture Rates csv file
recycling = pd.read_csv("~/Desktop/Springboard/Capstone Project 1/Data/Recycling_Diversion_and_Capture_Rates (2).csv")
recycling.head()

Zone District  Fiscal Month Number  Fiscal Year Month Name  \
0  Brooklyn North    BKN01                   10         2019      April   
1  Brooklyn North    BKN02                   10         2019      April   
2  Brooklyn North    BKN03                   10         2019      April   
3  Brooklyn North    BKN04                   10         2019      April   
4  Brooklyn North    BKN05                   10         2019      April   

   Diversion Rate-Total (Total Recycling / Total Waste)  \
0                                               14.7      
1                                               20.0      
2                                               12.2      
3                                               15.5      
4                                               10.1      

   Capture Rate-Paper (Total Paper / Max Paper)  \
0                                          44.9   
1                                          34.2   
2                                          33.5   
3                                          35.2   
4                                          22.3   

   Capture Rate-MGP (Total MGP / Max MGP)  \
0                                    43.0   
1                                    57.9   
2                                    44.9   
3                                    68.5   
4                                    45.1   

   Capture Rate-Total ((Total Recycling - Leaves (Recycling)) / (Max Paper + Max MGP))x100  
0                                               44.1                                        
1                                               41.2                                        
2                                               38.2                                        
3                                               48.8                                        
4                                               31.5

Above is the preview of the first 5 rows of the dataset, and we can see that there are 9 columns that are all shown including Zone, District, and 4 different rates with their definitions. Since the head attribute of a pandas dataframe only displays a specified number of rows (default of 5) and a limited number of columns that can fit in the window, we will create a function that can allow us to get a quick glimpse of the total rows and columns in a particular dataframe as we work through the wrangling process.

In [4]:
#Function to print the shape of 
def shape_df(df):
    print("This dataframe has {r} rows and {c} columns.".format(r = df.shape[0], c = df.shape[1]))

shape_df(recycling)    

This dataframe has 2832 rows and 9 columns.


Above we define the simple function which prints the rows and columns of a dataframe using the Pandas built in shape attributes. When calling the function with the our recycling dataframe as the input, we discover it has a total of 2832 rows and 9 columns. We can continue to use this function to verify the updates in size of each subsequent wrangled dataframe.

Now that we have a preview of the recycling dataframe, we will load the community district indicators data CSV file the same way using the Pandas read_csv function, and store this in the variable 'districts_raw'.

In [5]:
#Read in and Preview Community District Indicators Data 
districts_raw = pd.read_csv("/Users/varunraja/Desktop/Springboard/Capstone Project 1/Data/Bronx-7-indicators.csv")
districts_raw.head()

the_geom  lots_parking  the_geom_webmercator   acres  \
0       NaN            23                   NaN   976.3   
1       NaN            52                   NaN   865.9   
2       NaN            57                   NaN  1076.9   
3       NaN            77                   NaN  1131.6   
4       NaN            39                   NaN  1005.4   

                                         acs_tooltip  \
0  American Community Survey 2013-2017 5-Year Est...   
1  American Community Survey 2013-2017 5-Year Est...   
2  American Community Survey 2013-2017 5-Year Est...   
3  American Community Survey 2013-2017 5-Year Est...   
4  American Community Survey 2013-2017 5-Year Est...   

                                       acs_tooltip_2  \
0  American Community Survey (ACS) 2013-2017 5-ye...   
1  American Community Survey (ACS) 2013-2017 5-ye...   
2  American Community Survey (ACS) 2013-2017 5-ye...   
3  American Community Survey (ACS) 2013-2017 5-ye...   
4  American Community Survey (ACS) 2013-2017 5-ye...   

                                       acs_tooltip_3  area_sqmi  borocd  \
0  2010 Census population counts for floodplain a...        1.5     101   
1  2010 Census population counts for floodplain a...        1.4     102   
2  2010 Census population counts for floodplain a...        1.7     103   
3  2010 Census population counts for floodplain a...        1.8     104   
4  2010 Census population counts for floodplain a...        1.6     105   

                cb_email  ...         son_issue_1  \
0       man01@cb.nyc.gov  ...          Resiliency   
1    bgormley@cb.nyc.gov  ...  Affordable housing   
2  info@cb3manhattan.org  ...  Affordable housing   
3     jbodine@cb.nyc.gov  ...  Affordable housing   
4         office@cb5.org  ...  Affordable housing   

                                         son_issue_2  \
0                                            Traffic   
1                                              Parks   
2                                    Senior services   
3     Neighborhood preservation (development trends)   
4  Social services (including services for the ho...   

                 son_issue_3 total_lot_area  under18_rate  under18_rate_boro  \
0                      Other       36115322          12.8               14.5   
1                    Schools       25696432          12.8               14.5   
2                      Other       33692052          11.7               14.5   
3                    Traffic       38177268           8.2               14.5   
4  Transit (buses & subways)       28192622           8.2               14.5   

   under18_rate_nyc  unemployment_boro  unemployment_cd  unemployment_nyc  
0                21                4.2              2.8               4.9  
1                21                4.2              2.8               4.9  
2                21                4.2              4.1               4.9  
3                21                4.2              3.7               4.9  
4                21                4.2              3.7               4.9  

[5 rows x 184 columns]

In [6]:
shape_df(districts_raw)    

This dataframe has 59 rows and 184 columns.


As shown above by the head preview of the District indicator file, this dataframe has several more columns, some of which are hidden by the limited space. We can use the previously defined function to get a count of the rows and columns and finx it to be 59 x 184. 

Since we know there are 184 columns but still cannot see each column in the preview, we could discover this by using the pd.columns function to print all the column headers to get a better idea of the data.

In [7]:
districts_raw.columns.values

array(['the_geom', 'lots_parking', 'the_geom_webmercator', 'acres',
       'acs_tooltip', 'acs_tooltip_2', 'acs_tooltip_3', 'area_sqmi',
       'borocd', 'cb_email', 'cb_website', 'cd_full_title',
       'cd_short_title', 'cd_son_fy2018', 'cd_tot_bldgs',
       'cd_tot_resunits', 'count_hosp_clinic', 'count_libraries',
       'count_parks', 'count_public_schools', 'crime_count',
       'crime_count_boro', 'crime_count_nyc', 'crime_per_1000',
       'crime_per_1000_boro', 'crime_per_1000_nyc', 'female_10_14',
       'female_15_19', 'female_20_24', 'female_25_29', 'female_30_34',
       'female_35_39', 'female_40_44', 'female_45_49', 'female_5_9',
       'female_50_54', 'female_55_59', 'female_60_64', 'female_65_69',
       'female_70_74', 'female_75_79', 'female_80_84', 'female_85_over',
       'female_under_5', 'fp_100_area', 'fp_100_bldg',
       'fp_100_cost_burden', 'fp_100_cost_burden_value', 'fp_100_mhhi',
       'fp_100_mortg_value', 'fp_100_openspace', 'fp_100_ownerocc',
       

Taking a look at all of the the columns in the dataframe and using the data dictionary to understand the what each column records, it's evident that not every column will be useful for analysis and many may be completely irrelevant altogether. As such, we can clean up this dataframe by only selecting the columns we need and creating a new dataframe with fewer columns that will be easier to work with.

In [8]:
#Update community indicators dataframe with only specified columns that are useful - create new dataframe called 'districts_filtered' with these
indicator_columns = ['cd_short_title', 'puma', 'area_sqmi', 'pop_2010', 'poverty_rate', 'pct_bach_deg', 'unemployment_cd', 'mean_commute', 'pct_hh_rent_burd', 'pct_clean_strts', 'crime_per_1000', 'count_parks', 'count_hosp_clinic', 'count_libraries', 'count_public_schools', 'pct_foreign_born', 'lep_rate', 'under18_rate', 'male_under_5', 'female_under_5', 'male_5_9', 'female_5_9', 'male_10_14', 'female_10_14', 'male_15_19', 'female_15_19', 'male_20_24', 'female_20_24', 'male_25_29', 'female_25_29', 'male_30_34', 'female_30_34', 'male_35_39', 'female_35_39', 'male_40_44', 'female_40_44', 'male_45_49', 'female_45_49', 'male_50_54', 'female_50_54', 'male_55_59', 'female_55_59', 'male_60_64', 'female_60_64', 'male_65_69', 'female_65_69', 'male_70_74', 'female_70_74', 'male_75_79', 'female_75_79', 'male_80_84', 'female_80_84', 'male_85_over', 'female_85_over', 'pct_white_nh', 'pct_black_nh', 'pct_asian_nh', 'pct_other_nh', 'pct_hispanic', 'fp_100_pop']
districts = districts_raw.loc[:, indicator_columns]
shape_df(districts)

This dataframe has 59 rows and 60 columns.


As shown above, we've narrowed the dataframe down to 59 rows, accounting for each of NYC 59 community districts, as well as 60 columns for any relevant information on district name, borough, and the measurement for that district. Since many of the columns are quantities and percentages, another step to take that will ensure that all of them add up to 100% is to fill in any blank values and replace them with zero. Upon reviewing the dataframe, it looks like there are a handful of these and when adding up the percentages would make them zero.

In [9]:
#Replace any blank values in dataframe with zero as dataset indicates these are zero based on percentages adding up to 100
districts_raw.fillna(0, inplace=True)

After addressing the missing and blank values, we have a dataframe with data filled in all the rows and columns necessary. Going back and reviewing the columns of district level data, there are two things that stand out regarding percentages of different subgroups. First, while there is a column for each increment of age group roughly every five years for both male and female, there does not seem to be one indicating total percentage of male and female across all age groups. As such, we can easily add this by creating two new columns for 'female_total' and 'male_total' that are the sum of the percentages of all the male and female columns, as shown below.

In [10]:
#Create column with percentage of Male and Female by summing the total percentages of Males and Females across all age groups
districts['female_total'] = districts['female_under_5'] + districts['female_5_9'] + districts['female_10_14'] + districts['female_15_19'] + districts['female_20_24'] + districts['female_25_29'] + districts['female_30_34'] + districts['female_35_39'] + districts['female_40_44'] + districts['female_45_49'] + districts['female_50_54'] + districts['female_55_59'] + districts['female_60_64'] + districts['female_65_69'] + districts['female_70_74'] + districts['female_75_79'] + districts['female_80_84'] + districts['female_85_over']
districts['male_total'] = districts['male_under_5'] + districts['male_5_9'] + districts['male_10_14'] + districts['male_15_19'] + districts['male_20_24'] + districts['male_25_29'] + districts['male_30_34'] + districts['male_35_39'] + districts['male_40_44'] + districts['male_45_49'] + districts['male_50_54'] + districts['male_55_59'] + districts['male_60_64'] + districts['male_65_69'] + districts['male_70_74'] + districts['male_75_79'] + districts['male_80_84'] + districts['male_85_over']

After aggregating the male and female total percentages across age groups into newly added columns, another similar exercise that would be useful would be to aggregate the age group percentages. Since the raw data has them broken down into over a dozen groups for every 5 years, there is less potential for a lot more variance and less conclusions to be made from any high correlation of one age group. 

As such, it would be beneficial to aggregate these into larger age group ranges since it's more likely to draw conclusions from these. Below we'll add 4 new columns for wider age groups of each 20 years, and define them as minors (under 20), young adults (20-39, middle aged (40-59), and seniors (60+). While the range of 20 years is somewhat arbitrary, this breakdown is more common in defining generational gap differences. 

In [11]:
#Create column with percentage of Age groups (0-20, 20-40, 40-59, 60+) by summing age groups across Male and Female classifiers
districts['minors_under20'] = districts['female_under_5'] + districts['female_5_9'] + districts['female_10_14'] + districts['female_15_19'] + districts['male_under_5'] + districts['male_5_9'] + districts['male_10_14'] + districts['male_15_19'] 
districts['youngadults_20-39'] = districts['female_20_24'] + districts['female_25_29'] + districts['female_30_34'] + districts['female_35_39'] + districts['male_20_24'] + districts['male_25_29'] + districts['male_30_34'] + districts['male_35_39']
districts['middleaged_40-59'] = districts['female_40_44'] + districts['female_45_49'] + districts['female_50_54'] + districts['female_55_59'] + districts['male_40_44'] + districts['male_45_49'] + districts['male_50_54'] + districts['male_55_59']
districts['senior_over60'] = districts['female_60_64'] + districts['female_65_69'] + districts['female_70_74'] + districts['female_75_79'] + districts['female_80_84'] + districts['female_85_over'] + districts['male_60_64'] + districts['male_65_69'] + districts['male_70_74'] + districts['male_75_79'] + districts['male_80_84'] + districts['male_85_over']

After creating the aggregated age groups and gender columns along with filling in missing values, the districts dataframe is essentially complete with all the data we need from it. 

The next major component to the wrangling process is to combine the districts data with the recycling data. Taking a look at the 'District' column in the 'recycling' dataframe and the 'cd_short_title' column in the 'districts', we see that they both indicate the community district name as the borough name combined with a number. However, the fields are not an exact match. For example, the recycling dataframe lists Brooklyn Zone 1 as 'BKN01' and the districts dataframe lists the same as 'Brooklyn CD 1'. As such, we need to either create a key to map the two to matching keywords so that we can join the two dataframes together.

To do this, we start by creating a data dictionary for both the recycling and districts dataframes that matches the district name in their dataset to a standard name (for our purposes we'll use the format 'Borough-District Number').

In [12]:
#Standardize the Community District Keys in both the 'recycling' dataframe and 'districts_filtered' dataframe in order to merge on new key
#First create dictionaries to match values in each dataframe with a standardized key that will later be used to join
recycling_dict = {'BKN01':'Brooklyn-01', 'BKN02':'Brooklyn-02', 'BKN03':'Brooklyn-03', 'BKN04':'Brooklyn-04', 'BKN05':'Brooklyn-05', 'BKS06':'Brooklyn-06', 'BKS07':'Brooklyn-07', 'BKN08':'Brooklyn-08', 'BKN09':'Brooklyn-09', 'BKS10':'Brooklyn-10', 'BKS11':'Brooklyn-11', 'BKS12':'Brooklyn-12', 'BKS13':'Brooklyn-13', 'BKS14':'Brooklyn-14', 'BKS15':'Brooklyn-15', 'BKN16':'Brooklyn-16', 'BKN17':'Brooklyn-17', 'BKS18':'Brooklyn-18', 'BX01':'Bronx-01', 'BX02':'Bronx-02', 'BX03':'Bronx-03', 'BX04':'Bronx-04', 'BX05':'Bronx-05', 'BX06':'Bronx-06', 'BX07':'Bronx-07', 'BX08':'Bronx-08', 'BX09':'Bronx-09', 'BX10':'Bronx-10', 'BX11':'Bronx 11', 'BX12':'Bronx 12', 'MN01':'Manhattan-01', 'MN02':'Manhattan-02', 'MN03':'Manhattan-03', 'MN04':'Manhattan-04', 'MN05':'Manhattan-05', 'MN06':'Manhattan-06', 'MN07':'Manhattan-07', 'MN08':'Manhattan-08', 'MN09':'Manhattan-09', 'MN10':'Manhattan-10', 'MN11':'Manhattan-11', 'MN12':'Manhattan-12', 'QW01':'Queens-01', 'QW02':'Queens-02', 'QW03':'Queens-03', 'QW04':'Queens-04', 'QW05':'Queens-05', 'QW06':'Queens-06', 'QE07':'Queens-07', 'QE08':'Queens-08', 'QW09':'Queens-09', 'QE10':'Queens-10', 'QE11':'Queens-11', 'QE12':'Queens-12', 'QE13':'Queens-13', 'QE14':'Queens-14', 'SI01':'Staten_Island-01', 'SI02':'Staten_Island-02', 'SI03':'Staten_Island-03'}
district_dict = {'Brooklyn CD 1':'Brooklyn-01', 'Brooklyn CD 2':'Brooklyn-02', 'Brooklyn CD 3':'Brooklyn-03', 'Brooklyn CD 4':'Brooklyn-04', 'Brooklyn CD 5':'Brooklyn-05', 'Brooklyn CD 6':'Brooklyn-06', 'Brooklyn CD 7':'Brooklyn-07', 'Brooklyn CD 8':'Brooklyn-08', 'Brooklyn CD 9':'Brooklyn-09', 'Brooklyn CD 10':'Brooklyn-10', 'Brooklyn CD 11':'Brooklyn-11', 'Brooklyn CD 12':'Brooklyn-12', 'Brooklyn CD 13':'Brooklyn-13', 'Brooklyn CD 14':'Brooklyn-14', 'Brooklyn CD 15':'Brooklyn-15', 'Brooklyn CD 16':'Brooklyn-16', 'Brooklyn CD 17':'Brooklyn-17', 'Brooklyn CD 18':'Brooklyn-18', 'Bronx CD 1':'Bronx-01', 'Bronx CD 2':'Bronx-02', 'Bronx CD 3':'Bronx-03', 'Bronx CD 4':'Bronx-04', 'Bronx CD 5':'Bronx-05', 'Bronx CD 6':'Bronx-06', 'Bronx CD 7':'Bronx-07', 'Bronx CD 8':'Bronx-08', 'Bronx CD 9':'Bronx-09', 'Bronx CD 10':'Bronx-10', 'Bronx CD 11':'Bronx 11', 'Bronx CD 12':'Bronx 12', 'Manhattan CD 1':'Manhattan-01', 'Manhattan CD 2':'Manhattan-02', 'Manhattan CD 3':'Manhattan-03', 'Manhattan CD 4':'Manhattan-04', 'Manhattan CD 5':'Manhattan-05', 'Manhattan CD 6':'Manhattan-06', 'Manhattan CD 7':'Manhattan-07', 'Manhattan CD 8':'Manhattan-08', 'Manhattan CD 9':'Manhattan-09', 'Manhattan CD 10':'Manhattan-10', 'Manhattan CD 11':'Manhattan-11', 'Manhattan CD 12':'Manhattan-12', 'Queens CD 1':'Queens-01', 'Queens CD 2':'Queens-02', 'Queens CD 3':'Queens-03', 'Queens CD 4':'Queens-04', 'Queens CD 5':'Queens-05', 'Queens CD 6':'Queens-06', 'Queens CD 7':'Queens-07', 'Queens CD 8':'Queens-08', 'Queens CD 9':'Queens-09', 'Queens CD 10':'Queens-10', 'Queens CD 11':'Queens-11', 'Queens CD 12':'Queens-12', 'Queens CD 13':'Queens-13', 'Queens CD 14':'Queens-14', 'Staten Island CD 1':'Staten_Island-01', 'Staten Island CD 2':'Staten_Island-02', 'Staten Island CD 3':'Staten_Island-03'}

After creating the dictionary and standardizing the community districts names, we can use the pandas method df.map to create a new column that maps the district name based on the dictionary, as shown below.

In [13]:
#Update dataframes with new column including standardized key to join called "District_Join"
recycling['District_Join']= recycling['District'].map(recycling_dict)
districts['District_Join']= districts['cd_short_title'].map(district_dict)

In [14]:
#Move new column "District_Join" to front of both dataframes for increased readability
recycling = recycling[['District_Join', 'Zone', 'District', 'Fiscal Month Number', 'Fiscal Year', 'Month Name',
       'Diversion Rate-Total (Total Recycling / Total Waste)',
       'Capture Rate-Paper (Total Paper / Max Paper)',
       'Capture Rate-MGP (Total MGP / Max MGP)',
       'Capture Rate-Total ((Total Recycling - Leaves (Recycling)) / (Max Paper + Max MGP))x100']]

districts = districts[['District_Join', 'cd_short_title', 'puma', 'area_sqmi', 'pop_2010', 'poverty_rate',
       'pct_bach_deg', 'unemployment_cd', 'mean_commute', 'pct_hh_rent_burd',
       'pct_clean_strts', 'crime_per_1000', 'count_parks', 'count_hosp_clinic',
       'count_libraries', 'count_public_schools', 'pct_foreign_born',
       'lep_rate', 'under18_rate', 'male_under_5', 'female_under_5',
       'male_5_9', 'female_5_9', 'male_10_14', 'female_10_14', 'male_15_19',
       'female_15_19', 'male_20_24', 'female_20_24', 'male_25_29',
       'female_25_29', 'male_30_34', 'female_30_34', 'male_35_39',
       'female_35_39', 'male_40_44', 'female_40_44', 'male_45_49',
       'female_45_49', 'male_50_54', 'female_50_54', 'male_55_59',
       'female_55_59', 'male_60_64', 'female_60_64', 'male_65_69',
       'female_65_69', 'male_70_74', 'female_70_74', 'male_75_79',
       'female_75_79', 'male_80_84', 'female_80_84', 'male_85_over',
       'female_85_over', 'pct_white_nh', 'pct_black_nh', 'pct_asian_nh',
       'pct_other_nh', 'pct_hispanic', 'fp_100_pop', 'female_total',
       'male_total', 'minors_under20', 'youngadults_20-39', 'middleaged_40-59',
       'senior_over60']]

Once we've mapped and rearranged the districts columns to where the two dataframes values match one another, we can do a simple merge by using the Pandas merge function and specifying the column to join as the new 'district_join' column created above through the dictionary mapping.

In [15]:
#Merge the Recyling Rate and District Info Tables
recycling_districts = pd.merge(recycling, districts, on='District_Join')
recycling_districts.head()

District_Join            Zone District  Fiscal Month Number  Fiscal Year  \
0   Brooklyn-01  Brooklyn North    BKN01                   10         2019   
1   Brooklyn-01  Brooklyn North    BKN01                    2         2019   
2   Brooklyn-01  Brooklyn North    BKN01                    6         2019   
3   Brooklyn-01  Brooklyn North    BKN01                    8         2019   
4   Brooklyn-01  Brooklyn North    BKN01                    7         2019   

  Month Name  Diversion Rate-Total (Total Recycling / Total Waste)  \
0      April                                               14.7      
1     August                                               16.6      
2   December                                               16.5      
3   February                                               15.9      
4    January                                               17.1      

   Capture Rate-Paper (Total Paper / Max Paper)  \
0                                          44.9   
1                                          51.2   
2                                          52.3   
3                                          48.3   
4                                          49.9   

   Capture Rate-MGP (Total MGP / Max MGP)  \
0                                    43.0   
1                                    44.8   
2                                    46.8   
3                                    48.2   
4                                    52.0   

   Capture Rate-Total ((Total Recycling - Leaves (Recycling)) / (Max Paper + Max MGP))x100  \
0                                               44.1                                         
1                                               48.3                                         
2                                               50.0                                         
3                                               48.2                                         
4                                               50.7                                         

   ... pct_asian_nh  pct_other_nh  pct_hispanic  fp_100_pop  female_total  \
0  ...          7.5           2.1          23.0        3397          50.8   
1  ...          7.5           2.1          23.0        3397          50.8   
2  ...          7.5           2.1          23.0        3397          50.8   
3  ...          7.5           2.1          23.0        3397          50.8   
4  ...          7.5           2.1          23.0        3397          50.8   

   male_total  minors_under20  youngadults_20-39  middleaged_40-59  \
0        49.4            22.9               44.5              18.8   
1        49.4            22.9               44.5              18.8   
2        49.4            22.9               44.5              18.8   
3        49.4            22.9               44.5              18.8   
4        49.4            22.9               44.5              18.8   

   senior_over60  
0           14.0  
1           14.0  
2           14.0  
3           14.0  
4           14.0  

[5 rows x 76 columns]

In [16]:
shape_df(recycling_districts)

This dataframe has 2832 rows and 76 columns.


After creating and previewing the dataframe, we see that it has successfully joined on the 'District_Join' column and has a total of 2832 rows and 76 columns, which makes sense as there is recycling data for each month and year on the recycling dataframe. However, since we do not have access to the district level demographic data changing through time the same way the recycling data is, there is not as much use in keeping that as a variable. 

As such, we can quickly simplify the combined dataframe further by aggregating the monthly recycling rates into annual data and perhaps even further by only extracting a single years data. We will start by combining the monthly data into annual from 2016-2019 by running the df.groupby method on the dataframe and aggregating by the mean recycling rates, while maintaining the districts, borough, and year.

In [17]:
#recycling_districts.to_excel('check.xlsx')

In [22]:
#Aggregate the years column to take the monthly rows for each community district per year and find the mean rates
df = recycling_districts.groupby(['District_Join', 'Fiscal Year', 'Zone'], as_index=False).aggregate('mean')
shape_df(df)
df.head()

This dataframe has 236 rows and 73 columns.


District_Join  Fiscal Year   Zone  Fiscal Month Number  \
0      Bronx 11         2016  Bronx                  6.5   
1      Bronx 11         2017  Bronx                  6.5   
2      Bronx 11         2018  Bronx                  6.5   
3      Bronx 11         2019  Bronx                  6.5   
4      Bronx 12         2016  Bronx                  6.5   

   Diversion Rate-Total (Total Recycling / Total Waste)  \
0                                          14.166667      
1                                          14.200000      
2                                          16.916667      
3                                          17.025000      
4                                          16.175000      

   Capture Rate-Paper (Total Paper / Max Paper)  \
0                                     35.033333   
1                                     33.341667   
2                                     34.558333   
3                                     34.116667   
4                                     33.950000   

   Capture Rate-MGP (Total MGP / Max MGP)  \
0                               53.450000   
1                               56.375000   
2                               57.133333   
3                               59.341667   
4                               71.833333   

   Capture Rate-Total ((Total Recycling - Leaves (Recycling)) / (Max Paper + Max MGP))x100  \
0                                          41.750000                                         
1                                          41.725000                                         
2                                          42.791667                                         
3                                          43.333333                                         
4                                          47.808333                                         

     puma  area_sqmi  ...  pct_asian_nh  pct_other_nh  pct_hispanic  \
0  3704.0        3.6  ...           8.0           2.2          46.7   
1  3704.0        3.6  ...           8.0           2.2          46.7   
2  3704.0        3.6  ...           8.0           2.2          46.7   
3  3704.0        3.6  ...           8.0           2.2          46.7   
4  3702.0        5.6  ...           3.1           2.3          22.9   

   fp_100_pop  female_total  male_total  minors_under20  youngadults_20-39  \
0       296.0          52.0        48.1            24.8               29.8   
1       296.0          52.0        48.1            24.8               29.8   
2       296.0          52.0        48.1            24.8               29.8   
3       296.0          52.0        48.1            24.8               29.8   
4        40.0          54.4        45.9            25.2               29.0   

   middleaged_40-59  senior_over60  
0              25.9           19.6  
1              25.9           19.6  
2              25.9           19.6  
3              25.9           19.6  
4              27.5           18.6  

[5 rows x 73 columns]

We now have a dataframe condensed into 236 rows, which is the data for 59 districts in each of the 4 fiscal years from 2016-2019, and the same 73 columns as earlier. As seen by the recycling rates, they are now the mean value of each month in a fiscal year.

While having 4 years of data for the recycling rates will be useful in order to see how the rates have changed over time at a high level, the district level demographic data still does not account for changes through time. As such, we will create a new dataframe to isolate one year's recycling data. Since most of the fields of district level data are measured as 5-year estimates from 2012-2016, we will use the year 2016 since it's likely to be closest to representing the 5-year estimates.

In [24]:
#Create new dataframe out of recycling data from fiscal year 2016 since it is closest to the time-frame of districts data
df_2016 = df[df['Fiscal Year'] == 2016]
shape_df(df_2016)
df_2016.head()

This dataframe has 59 rows and 73 columns.


District_Join  Fiscal Year   Zone  Fiscal Month Number  \
0       Bronx 11         2016  Bronx                  6.5   
4       Bronx 12         2016  Bronx                  6.5   
8       Bronx-01         2016  Bronx                  6.5   
12      Bronx-02         2016  Bronx                  6.5   
16      Bronx-03         2016  Bronx                  6.5   

    Diversion Rate-Total (Total Recycling / Total Waste)  \
0                                           14.166667      
4                                           16.175000      
8                                            6.450000      
12                                          12.225000      
16                                           9.266667      

    Capture Rate-Paper (Total Paper / Max Paper)  \
0                                      35.033333   
4                                      33.950000   
8                                      18.866667   
12                                     32.050000   
16                                     25.683333   

    Capture Rate-MGP (Total MGP / Max MGP)  \
0                                53.450000   
4                                71.833333   
8                                23.850000   
12                               49.800000   
16                               36.125000   

    Capture Rate-Total ((Total Recycling - Leaves (Recycling)) / (Max Paper + Max MGP))x100  \
0                                           41.750000                                         
4                                           47.808333                                         
8                                           21.058333                                         
12                                          39.850000                                         
16                                          30.283333                                         

      puma  area_sqmi  ...  pct_asian_nh  pct_other_nh  pct_hispanic  \
0   3704.0        3.6  ...           8.0           2.2          46.7   
4   3702.0        5.6  ...           3.1           2.3          22.9   
8   3710.0        2.2  ...           0.6           1.2          67.5   
12  3710.0        2.2  ...           0.6           1.2          67.5   
16  3705.0        1.6  ...           0.9           1.1          63.3   

    fp_100_pop  female_total  male_total  minors_under20  youngadults_20-39  \
0        296.0          52.0        48.1            24.8               29.8   
4         40.0          54.4        45.9            25.2               29.0   
8        580.0          50.2        49.7            31.6               33.3   
12       848.0          50.2        49.7            31.6               33.3   
16         0.0          53.6        46.6            32.8               31.2   

    middleaged_40-59  senior_over60  
0               25.9           19.6  
4               27.5           18.6  
8               22.7           12.3  
12              22.7           12.3  
16              23.3           12.9  

[5 rows x 73 columns]

Although we eliminated the monthly data, since it is available we can still use it to at least see the change in recycling rates over time. As such, we will create a separate dataframe for this and convert the months to numeric to be graphed below.

In [24]:
#Aggregate monthly data across boroughs in order to graph data across time in data story
df_month = recycling_districts.groupby(['Fiscal Month Number', 'Fiscal Year'], as_index=False).aggregate('mean')
df_month['FullDate'] = df_month['Fiscal Month Number'].astype(str) + df_month['Fiscal Year'].astype(str)
df_month['Fiscal Month Number'] = pd.to_numeric(df_month['Fiscal Month Number'])
df_month['Fiscal Year'] = pd.to_numeric(df_month['Fiscal Year'])
df_month['FullDate'] = pd.to_numeric(df_month['FullDate'])
df_month['Index'] = range(0, len(df_month))

Now that we have the 2016 combined dataframe, annual dataframe, and monthly, the datasets are more or less clean enough to work with for data analysis. There are however a few minor tweaks to add that will simplify the data visualization process and make it cleaner.

One of these is to combine the 'Zone' column labels for subdivisions with NYC boroughs. As seen by the data, there are two different divisions of both Brooklyn (Norht and South) and Queens (East and West). In order to simplify the data and look at in a borough level, we will combine these into one.

The other step is to delete any unnecessary columns - one that stands out is 'Fiscal Month Number' which was negated as a result of aggregating the monthly columns into annual.

In [25]:
%%capture
#Combine Brooklyn and Queens split Zones into one in order to later visualize across boroughs
df['Zone'].replace({'Brooklyn North': 'Brooklyn', 'Brooklyn South': 'Brooklyn', 'Queens West': 'Queens', 'Queens East': 'Queens'}, inplace=True)
df_2016['Zone'].replace({'Brooklyn North': 'Brooklyn', 'Brooklyn South': 'Brooklyn', 'Queens West': 'Queens', 'Queens East': 'Queens'}, inplace=True)

In [26]:
#Delete unnessecary columns - Fiscal Month Number
del df['Fiscal Month Number']
del df_2016['Fiscal Month Number']

In [27]:
shape_df(df), shape_df(df_2016)

This dataframe has 236 rows and 72 columns.
This dataframe has 59 rows and 72 columns.


(None, None)

Finally we are left with with finished dataframes with the rows and columns relevant to them. The last step we will take in order to continue the analysis in the Data Story & Statistical analysis as well as Machine Learning section is to store each of these dataframes in CSV files so they can be loaded into those notebooks, as shown below.

In [25]:
#Save dataframes as CSV in order to use in Data Story and ML notebooks
df_2016.to_csv('recycling_districts_2016')
df.to_csv('recycling_districts_all')
df_month.to_csv('recycling_districts_monthly')